The job `mingw-check` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=29291&view=logs&jobId=74054d28-c774-5fab-ecc2-352fe71b230e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/29291/logs/71), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/29291/logs/71)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[section]Starting: Linux mingw-check
##[section]Starting: Initialize job
Agent name: 'Azure Pipelines 43'
Agent machine name: 'fv-az578'
Current agent version: '2.168.2'
##[group]Operating System
16.04.6
LTS
LTS
##[endgroup]
##[group]Virtual Environment
Environment: ubuntu-16.04
Version: 20200430.2
Included Software: https://github.com/actions/virtual-environments/blob/ubuntu16/20200430.2/images/linux/Ubuntu1604-README.md
##[endgroup]
Agent running as: 'vsts'
Prepare build directory.
Set build variables.
Download all required tasks.
Download all required tasks.
Downloading task: Bash (3.163.2)
Checking job knob settings.
   Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
   Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
Start tracking orphan processes.
##[section]Finishing: Initialize job
##[section]Starting: Configure Job Name
==============================================================================
---
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/610cc602-0c37-4337-96aa-5f643b8dcd64.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/72239/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/72239/merge:refs/remotes/pull/72239/merge
---
 ---> 3adb0605cc65
Step 6/7 : ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
 ---> Using cache
 ---> 28dbc326cb7f
Step 7/7 : ENV SCRIPT python3 ../x.py test src/tools/expand-yaml-anchors &&            python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu &&            python3 ../x.py build --stage 0 src/tools/build-manifest &&            python3 ../x.py test --stage 0 src/tools/compiletest &&            python3 ../x.py test src/tools/tidy &&            python3 ../x.py doc --stage 0 src/libstd &&            /scripts/validate-toolstate.sh
 ---> 537a01811900
Successfully built 537a01811900
Successfully tagged rust-ci:latest
Built container sha256:537a018119009dc218456238dec90b5530050db1e2a1e166550c218003f6159d
---
    Checking panic_unwind v0.0.0 (/checkout/src/libpanic_unwind)
error[E0053]: method `eq` has an incompatible type for trait
   --> src/libstd/net/addr.rs:677:5
    |
677 |     fn eq(&self, other: &SocketAddrV4) -> bool {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `net::addr::SocketAddr`, found struct `net::addr::SocketAddrV4`
    |
    = note: expected fn pointer `fn(&net::addr::SocketAddrV4, &net::addr::SocketAddr) -> _`
               found fn pointer `fn(&net::addr::SocketAddrV4, &net::addr::SocketAddrV4) -> _`
error[E0053]: method `eq` has an incompatible type for trait
   --> src/libstd/net/addr.rs:686:5
    |
    |
686 |     fn eq(&self, other: &SocketAddrV6) -> bool {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `net::addr::SocketAddr`, found struct `net::addr::SocketAddrV6`
    |
    = note: expected fn pointer `fn(&net::addr::SocketAddrV6, &net::addr::SocketAddr) -> _`
               found fn pointer `fn(&net::addr::SocketAddrV6, &net::addr::SocketAddrV6) -> _`
error[E0308]: mismatched types
   --> src/libstd/net/addr.rs:679:13
    |
678 |         match other {
678 |         match other {
    |               ----- this expression has type `&net::addr::SocketAddrV4`
679 |             SocketAddr::V4(v4) => self == v4,
    |             ^^^^^^^^^^^^^^^^^^ expected struct `net::addr::SocketAddrV4`, found enum `net::addr::SocketAddr`
error[E0308]: mismatched types
   --> src/libstd/net/addr.rs:680:13
    |
678 |         match other {
678 |         match other {
    |               ----- this expression has type `&net::addr::SocketAddrV4`
679 |             SocketAddr::V4(v4) => self == v4,
680 |             SocketAddr::V6(_) => false,
    |             ^^^^^^^^^^^^^^^^^ expected struct `net::addr::SocketAddrV4`, found enum `net::addr::SocketAddr`
error[E0308]: mismatched types
   --> src/libstd/net/addr.rs:688:13
    |
687 |         match other {
687 |         match other {
    |               ----- this expression has type `&net::addr::SocketAddrV6`
688 |             SocketAddr::V4(_) => false,
    |             ^^^^^^^^^^^^^^^^^ expected struct `net::addr::SocketAddrV6`, found enum `net::addr::SocketAddr`
error[E0308]: mismatched types
   --> src/libstd/net/addr.rs:689:13
    |
687 |         match other {
687 |         match other {
    |               ----- this expression has type `&net::addr::SocketAddrV6`
688 |             SocketAddr::V4(_) => false,
689 |             SocketAddr::V6(v6) => self == v6,
    |             ^^^^^^^^^^^^^^^^^^ expected struct `net::addr::SocketAddrV6`, found enum `net::addr::SocketAddr`
error: aborting due to 6 previous errors

Some errors have detailed explanations: E0053, E0308.
For more information about an error, try `rustc --explain E0053`.
---
  local time: Fri May 15 19:04:22 UTC 2020
  network time: Fri, 15 May 2020 19:04:22 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/72239/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/72239/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (3383) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

--- a/src/libstd/net/addr.rs
+++ b/src/libstd/net/addr.rs
@@ -1,3 +1,4 @@
+use crate::cmp::Ordering;
 use crate::convert::TryInto;
 use crate::fmt;
 use crate::hash;
@@ -36,7 +37,7 @@ use crate::vec;
 /// assert_eq!(socket.port(), 8080);
 /// assert_eq!(socket.is_ipv4(), true);
 /// ```
-#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug, PartialOrd, Ord)]
 #[stable(feature = "rust1", since = "1.0.0")]
 pub enum SocketAddr {
     /// An IPv4 socket address.
@@ -653,10 +654,114 @@ impl PartialEq for SocketAddrV6 {
             && self.inner.sin6_scope_id == other.inner.sin6_scope_id
     }
 }
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl PartialEq<SocketAddrV4> for SocketAddr {
+    fn eq(&self, other: &SocketAddrV4) -> bool {
+        match self {
+            SocketAddr::V4(v4) => v4 == other,
+            SocketAddr::V6(_) => false,
+        }
+    }
+}
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl PartialEq<SocketAddrV6> for SocketAddr {
+    fn eq(&self, other: &SocketAddrV6) -> bool {
+        match self {
+            SocketAddr::V4(_) => false,
+            SocketAddr::V6(v6) => v6 == other,
+        }
+    }
+}
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl PartialEq<SocketAddr> for SocketAddrV4 {
+    fn eq(&self, other: &SocketAddr) -> bool {
+        match other {
+            SocketAddr::V4(v4) => self == v4,
+            SocketAddr::V6(_) => false,
+        }
+    }
+}
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl PartialEq<SocketAddr> for SocketAddrV6 {
+    fn eq(&self, other: &SocketAddr) -> bool {
+        match other {
+            SocketAddr::V4(_) => false,
+            SocketAddr::V6(v6) => self == v6,
+        }
+    }
+}
 #[stable(feature = "rust1", since = "1.0.0")]
 impl Eq for SocketAddrV4 {}
 #[stable(feature = "rust1", since = "1.0.0")]
 impl Eq for SocketAddrV6 {}
+
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl PartialOrd for SocketAddrV4 {
+    fn partial_cmp(&self, other: &SocketAddrV4) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl PartialOrd<SocketAddr> for SocketAddrV4 {
+    fn partial_cmp(&self, other: &SocketAddr) -> Option<Ordering> {
+        match other {
+            SocketAddr::V4(v4) => self.partial_cmp(v4),
+            SocketAddr::V6(_) => Some(Ordering::Less),
+        }
+    }
+}
+
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl PartialOrd<SocketAddrV4> for SocketAddr {
+    fn partial_cmp(&self, other: &SocketAddrV4) -> Option<Ordering> {
+        match self {
+            SocketAddr::V4(v4) => v4.partial_cmp(other),
+            SocketAddr::V6(_) => Some(Ordering::Greater),
+        }
+    }
+}
+
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl PartialOrd for SocketAddrV6 {
+    fn partial_cmp(&self, other: &SocketAddrV6) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl PartialOrd<SocketAddr> for SocketAddrV6 {
+    fn partial_cmp(&self, other: &SocketAddr) -> Option<Ordering> {
+        match other {
+            SocketAddr::V4(_) => Some(Ordering::Greater),
+            SocketAddr::V6(v6) => self.partial_cmp(v6),
+        }
+    }
+}
+
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl PartialOrd<SocketAddrV6> for SocketAddr {
+    fn partial_cmp(&self, other: &SocketAddrV6) -> Option<Ordering> {
+        match self {
+            SocketAddr::V4(_) => Some(Ordering::Less),
+            SocketAddr::V6(v6) => v6.partial_cmp(other),
+        }
+    }
+}
+
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl Ord for SocketAddrV4 {
+    fn cmp(&self, other: &SocketAddrV4) -> Ordering {
+        self.ip().cmp(other.ip()).then(self.port().cmp(&other.port()))
+    }
+}
+
+#[stable(feature = "socketaddr_ordering", since = "1.45.0")]
+impl Ord for SocketAddrV6 {
+    fn cmp(&self, other: &SocketAddrV6) -> Ordering {
+        self.ip().cmp(other.ip()).then(self.port().cmp(&other.port()))
+    }
+}
 
 #[stable(feature = "rust1", since = "1.0.0")]
 impl hash::Hash for SocketAddrV4 {
@@ -1101,5 +1206,45 @@ mod tests {
         ));
         assert!(!v6.is_ipv4());
         assert!(v6.is_ipv6());
+    }
+
+    #[test]
+    fn compare() {
+        let v4_1 = "224.120.45.1:23456".parse::<SocketAddrV4>().unwrap();
+        let v4_2 = "224.210.103.5:12345".parse::<SocketAddrV4>().unwrap();
+        let v4_3 = "224.210.103.5:23456".parse::<SocketAddrV4>().unwrap();
+        let v6_1 = "[2001:db8:f00::1002]:1234".parse::<SocketAddrV6>().unwrap();
+        let v6_2 = "[2001:db8:f00::2001]:1234".parse::<SocketAddrV6>().unwrap();
+        let v6_3 = "[2001:db8:f00::2001]:2345".parse::<SocketAddrV6>().unwrap();
+
+        // equality
+        assert_eq!(v4_1, SocketAddr::V4(v4_1));
+        assert_eq!(v6_1, SocketAddr::V6(v6_1));
+        assert_eq!(SocketAddr::V4(v4_1), SocketAddr::V4(v4_1));
+        assert_eq!(SocketAddr::V6(v6_1), SocketAddr::V6(v6_1));
+        assert!(v4_1 != SocketAddr::V6(v6_1));
+        assert!(v6_1 != SocketAddr::V4(v4_1));
+        assert!(v4_1 != v4_2);
+        assert!(v6_1 != v6_2);
+
+        // compare different addresses
+        assert!(v4_1 < v4_2);
+        assert!(v4_1 < SocketAddr::V4(v4_2));
+        assert!(SocketAddr::V4(v4_1) < v4_2);
+        assert!(SocketAddr::V4(v4_1) < SocketAddr::V4(v4_2));
+        assert!(v6_1 < v6_2);
+        assert!(v6_1 < SocketAddr::V6(v6_2));
+        assert!(SocketAddr::V6(v6_1) < v6_2);
+        assert!(SocketAddr::V6(v6_1) < SocketAddr::V6(v6_2));
+
+        // compare the same address with different ports
+        assert!(v4_2 < v4_3);
+        assert!(v4_2 < SocketAddr::V4(v4_3));
+        assert!(SocketAddr::V4(v4_2) < v4_3);
+        assert!(SocketAddr::V4(v4_2) < SocketAddr::V4(v4_3));
+        assert!(v6_2 < v6_3);
+        assert!(v6_2 < SocketAddr::V6(v6_3));
+        assert!(SocketAddr::V6(v6_2) < v6_3);
+        assert!(SocketAddr::V6(v6_2) < SocketAddr::V6(v6_3));
     }
 }

